File: webauthn-server-core/src/main/scala/com/yubico/webauthn/data/AttestationData.scala
package com.yubico.webauthn.data

import java.security.interfaces.ECPublicKey

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonIgnore
import com.yubico.u2f.data.messages.key.util.U2fB64Encoding
import com.yubico.webauthn.util.WebAuthnCodecs

case class AttestationData private[data] (

  /**
    * The AAGUID of the authenticator.
    */
  @JsonIgnore
  aaguid: ArrayBuffer,

  /**
    * The ID of the attested credential.
    */
  @JsonIgnore
  credentialId: ArrayBuffer,

  /**
    * The ''credential public key'' encoded in COSE_Key format.
    *
    * @todo verify requirements https://www.w3.org/TR/webauthn/#sec-attestation-data
    */
  @JsonIgnore
  credentialPublicKey: ArrayBuffer

) {

  @JsonProperty("aaguid")
  def aaguidBase64: String = U2fB64Encoding.encode(aaguid.toArray)

  @JsonProperty("credentialId")
  def credentialIdBase64: String = U2fB64Encoding.encode(credentialId.toArray)

  @JsonProperty("credentialPublicKey")
  def credentialPublicKeyBase64: String = U2fB64Encoding.encode(credentialPublicKey.toArray)

  def parsedCredentialPublicKey: ECPublicKey = WebAuthnCodecs.importCoseP256PublicKey(credentialPublicKey)

}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/data/AttestationData.java
package com.yubico.webauthn.data;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

@Value
@Builder
public class AttestationData {

    /**
     * The AAGUID of the authenticator.
     */
    @NonNull
    private final ByteArray aaguid;

    /**
     * The ID of the attested credential.
     */
    @NonNull
    private final ByteArray credentialId;

    /**
     * The ''credential public key'' encoded in COSE_Key format.
     *
     * @todo verify requirements https://www.w3.org/TR/webauthn/#sec-attestation-data
     */
    @NonNull
    private final ByteArray credentialPublicKey;

    @JsonCreator
    private AttestationData(
        @NonNull @JsonProperty("aaguid") ByteArray aaguid,
        @NonNull @JsonProperty("credentialId") ByteArray credentialId,
        @NonNull @JsonProperty("credentialPublicKey") ByteArray credentialPublicKey
    ) {
        this.aaguid = aaguid;
        this.credentialId = credentialId;
        this.credentialPublicKey = credentialPublicKey;
    }

}
